plugins/woocommerce/src/Internal/ProductFilters/TaxonomyHierarchyData.php

```diff
@@ -0,0 +1,285 @@
+			$term_id   = $term->term_id;
+			$parent_id = $term->parent;
+
+			$map['parents'][ $term_id ] = $parent_id;
+
+			if ( ! isset( $map['children'][ $parent_id ] ) ) {
+				$map['children'][ $parent_id ] = array();
+			}
+			$map['children'][ $parent_id ][] = $term_id;
+		}
+
+		// Pre-compute all descendants for each term.
+		foreach ( array_keys( $map['parents'] ) as $term_id ) {
+			$map['descendants'][ $term_id ] = $this->compute_descendants( $term_id, $map['children'] );
+		}
+
+		return $map;
+	}
+
+	/**
+	 * Build adjacency list hierarchy map.
+	 *
+	 * @param string $taxonomy The taxonomy name.
+	 * @return array Adjacency list hierarchy map.
+	 */
+	private function build_adjacency_hierarchy_map( string $taxonomy ): array {
+		$terms = get_terms(
+			array(
+				'taxonomy'   => $taxonomy,
+				'hide_empty' => false,
+				'fields'     => 'all',
+			)
+		);
+
+		if ( is_wp_error( $terms ) || empty( $terms ) ) {
+			return array();
+		}
+
+		$map = array(
+			'children' => array(),
+			'parents'  => array(),
+		);
+
+		foreach ( $terms as $term ) {
+			$term_id   = $term->term_id;
+			$parent_id = $term->parent;
+
+			$map['parents'][ $term_id ] = $parent_id;
+
+			if ( ! isset( $map['children'][ $parent_id ] ) ) {
+				$map['children'][ $parent_id ] = array();
+			}
+			$map['children'][ $parent_id ][] = $term_id;
+		}
+
+		return $map;
+	}
+
+	/**
+	 * Compute all descendants of a term.
+	 *
+	 * @param int   $term_id  The term ID.
+	 * @param array $children Children relationships map.
+	 * @return array Array of descendant term IDs.
+	 */
+	private function compute_descendants( int $term_id, array $children ): array {
+		$descendants = array();
+
+		if ( ! isset( $children[ $term_id ] ) ) {
+			return $descendants;
+		}
+
+		foreach ( $children[ $term_id ] as $child_id ) {
+			$descendants[] = $child_id;
+			$descendants   = array_merge( $descendants, $this->compute_descendants( $child_id, $children ) );
+		}
+
+		return array_unique( $descendants );
+	}
+
+	/**
+	 * Get parent term ID for a given term.
+	 *
+	 * @param int    $term_id  The term ID.
+	 * @param string $taxonomy The taxonomy name.
+	 * @return int The parent term ID (0 if root level).
+	 */
+	public function get_parent( int $term_id, string $taxonomy ): int {
+		$map = $this->get_hierarchy_map( $taxonomy );
+		return $map['parents'][ $term_id ] ?? 0;
+	}
+
+	/**
+	 * Get all descendants for a term.
+	 *
+	 * @param int    $term_id  The term ID.
+	 * @param string $taxonomy The taxonomy name.
+	 * @return array Array of all descendant term IDs.
+	 */
+	public function get_descendants( int $term_id, string $taxonomy ): array {
+		$map = $this->get_hierarchy_map( $taxonomy );
+
+		// Full map has pre-computed descendants.
+		if ( isset( $map['descendants'] ) ) {
+			return $map['descendants'][ $term_id ] ?? array();
+		}
+
+		// Adjacency map requires computation.
+		return $this->compute_descendants( $term_id, $map['children'] ?? array() );
```

Comment:
What about adding this to in-memory cache upon call?

```suggestion
		$descendants = $this->compute_descendants( $term_id, $map['children'] ?? array() );
		$map['descendants'][ $term_id ] = $descendants;
		return $descendants;
```